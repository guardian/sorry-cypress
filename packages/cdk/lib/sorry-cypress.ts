import { EndpointType } from '@aws-cdk/aws-apigateway';
import { Runtime } from '@aws-cdk/aws-lambda';
import type { App } from '@aws-cdk/core';
import { Duration } from '@aws-cdk/core';
import { GuApiLambda } from '@guardian/cdk';
import { Stage } from '@guardian/cdk/lib/constants';
import { GuCertificate } from '@guardian/cdk/lib/constructs/acm';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuCname } from '@guardian/cdk/lib/constructs/dns';

const app = 'sorry-cypress';

export class SorryCypress extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    // Domain name for the selected deployment stage
    const domainName = this.withStageDependentValue({
      app,
      variableName: 'domainName',
      stageValues: {
        [Stage.CODE]: 'sorry-cypress.code.dev-gutools.co.uk',
        [Stage.PROD]: 'sorry-cypress.gutools.co.uk',
      },
    });

    const certificate = new GuCertificate(this, {
      app,
      domainName,
    });

    const apiLambda = new GuApiLambda(this, 'lambda', {
      app,
      runtime: Runtime.NODEJS_14_X,
      apis: [
        {
          id: 'api',
          proxy: true,
          restApiName: `${app}-${this.stage}`,
          description: 'sorry-cypress server',
          endpointTypes: [EndpointType.REGIONAL],
          domainName: {
            certificate,
            domainName,
            endpointType: EndpointType.REGIONAL,
          },
        },
      ],
      fileName: 'sorry-cypress.zip',
      functionName: `${app}-lambda-api-${this.stage}`,
      handler: 'index.handler',
      monitoringConfiguration: { noMonitoring: true },
      memorySize: 256,
      timeout: Duration.seconds(15),
    });

    const maybeApi = apiLambda.apis.get('api');
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- we expect api to be available beyond this line
    const api = maybeApi!;

    // The custom domain name mapped to this API
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- we expect the domainName to be available here
    const apiDomain = api.domainName!;

    // CNAME mapping between API Gateway and the custom domain
    new GuCname(this, 'DNS entry', {
      app,
      domainName,
      ttl: Duration.hours(1),
      resourceRecord: apiDomain.domainNameAliasDomainName,
    });
  }
}
