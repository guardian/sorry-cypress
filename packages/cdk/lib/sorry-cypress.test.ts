import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { SorryCypress } from './sorry-cypress';

describe('The SorryCypress stack', () => {
  it('matches the snapshot', () => {
    const app = new App();
    const stack = new SorryCypress(app, 'SorryCypress', {
      stack: 'playground',
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
