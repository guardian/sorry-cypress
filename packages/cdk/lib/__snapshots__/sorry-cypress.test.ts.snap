// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SorryCypress stack matches the snapshot 1`] = `
Object {
  "Mappings": Object {
    "sorrycypress": Object {
      "CODE": Object {
        "domainName": "sorry-cypress.code.dev-gutools.co.uk",
      },
      "PROD": Object {
        "domainName": "sorry-cypress.gutools.co.uk",
      },
    },
  },
  "Outputs": Object {
    "lambdaapiEndpoint3B6C471A": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "lambdaapiC1812993",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "lambdaapiDeploymentStageprod9598BC2F",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertificateSorrycypress2A1EEDA2": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DomainName": Object {
          "Fn::FindInMap": Array [
            "sorrycypress",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "DNSentry": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::FindInMap": Array [
            "sorrycypress",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "lambdaapiCustomDomainEB2AF909",
              "RegionalDomainName",
            ],
          },
        ],
        "Stage": Object {
          "Ref": "Stage",
        },
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "lambda8B5974B5": Object {
      "DependsOn": Array [
        "lambdaServiceRoleDefaultPolicyBF6FA5E7",
        "lambdaServiceRole494E4CA6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                "playground/",
                Object {
                  "Ref": "Stage",
                },
                "/sorry-cypress/sorry-cypress.zip",
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "sorry-cypress",
            "STACK": "playground",
            "STAGE": Object {
              "Ref": "Stage",
            },
          },
        },
        "FunctionName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "sorry-cypress-lambda-api-",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "lambdaServiceRole494E4CA6",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaServiceRole494E4CA6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaServiceRoleDefaultPolicyBF6FA5E7": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/playground/sorry-cypress",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/playground/sorry-cypress/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaServiceRoleDefaultPolicyBF6FA5E7",
        "Roles": Array [
          Object {
            "Ref": "lambdaServiceRole494E4CA6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "lambdaapiANY4EBCADD1": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "lambda8B5974B5",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "lambdaapiC1812993",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "lambdaapiC1812993",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "lambdaapiANYApiPermissionSorryCypresslambdaapi63631BF5ANYEC64EA99": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "lambda8B5974B5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "lambdaapiC1812993",
              },
              "/",
              Object {
                "Ref": "lambdaapiDeploymentStageprod9598BC2F",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "lambdaapiANYApiPermissionTestSorryCypresslambdaapi63631BF5ANY8122E97D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "lambda8B5974B5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "lambdaapiC1812993",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "lambdaapiAccountAF0DCB01": Object {
      "DependsOn": Array [
        "lambdaapiC1812993",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "lambdaapiCloudWatchRole7E36513A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "lambdaapiC1812993": Object {
      "Properties": Object {
        "Description": "sorry-cypress server",
        "EndpointConfiguration": Object {
          "Types": Array [
            "REGIONAL",
          ],
        },
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              "sorry-cypress-",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "lambdaapiCloudWatchRole7E36513A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaapiCustomDomainEB2AF909": Object {
      "Properties": Object {
        "DomainName": Object {
          "Fn::FindInMap": Array [
            "sorrycypress",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "EndpointConfiguration": Object {
          "Types": Array [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": Object {
          "Ref": "CertificateSorrycypress2A1EEDA2",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "lambdaapiCustomDomainMapSorryCypresslambdaapi63631BF595056A2F": Object {
      "Properties": Object {
        "DomainName": Object {
          "Ref": "lambdaapiCustomDomainEB2AF909",
        },
        "RestApiId": Object {
          "Ref": "lambdaapiC1812993",
        },
        "Stage": Object {
          "Ref": "lambdaapiDeploymentStageprod9598BC2F",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "lambdaapiDeployment8E95B58501b308e2230dcb1addceb2c16e862385": Object {
      "DependsOn": Array [
        "lambdaapiproxyANYA94E968A",
        "lambdaapiproxyB573C729",
        "lambdaapiANY4EBCADD1",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "lambdaapiC1812993",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "lambdaapiDeploymentStageprod9598BC2F": Object {
      "DependsOn": Array [
        "lambdaapiAccountAF0DCB01",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "lambdaapiDeployment8E95B58501b308e2230dcb1addceb2c16e862385",
        },
        "RestApiId": Object {
          "Ref": "lambdaapiC1812993",
        },
        "StageName": "prod",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "sorry-cypress",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/sorry-cypress",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "lambdaapiproxyANYA94E968A": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "lambda8B5974B5",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "lambdaapiproxyB573C729",
        },
        "RestApiId": Object {
          "Ref": "lambdaapiC1812993",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "lambdaapiproxyANYApiPermissionSorryCypresslambdaapi63631BF5ANYproxyE32CEB2E": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "lambda8B5974B5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "lambdaapiC1812993",
              },
              "/",
              Object {
                "Ref": "lambdaapiDeploymentStageprod9598BC2F",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "lambdaapiproxyANYApiPermissionTestSorryCypresslambdaapi63631BF5ANYproxy828807EC": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "lambda8B5974B5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "lambdaapiC1812993",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "lambdaapiproxyB573C729": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "lambdaapiC1812993",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "lambdaapiC1812993",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
  },
}
`;
